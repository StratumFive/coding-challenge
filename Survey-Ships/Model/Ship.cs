using Survey_Ships.Helpers;

namespace Survey_Ships.Model
{
   public class Ship
   {
      public int PosX { get; set; }
      public int PosY { get; set; }
      public Direction Orientation { get; set; }

      public Ship(int posX, int posY, Direction orientation)
      {
         PosX = posX;
         PosY = posY;
         Orientation = orientation;
      }
   }
}
