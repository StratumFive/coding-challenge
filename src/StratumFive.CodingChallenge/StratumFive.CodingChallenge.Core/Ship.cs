using StratumFive.CodingChallenge.Core.Interfaces;
using System;

namespace StratumFive.CodingChallenge.Core
{
    public class Ship
    {
        public ShipState State { get; private set; } = ShipState.OK;

        public IntVector2 Position { get; private set; }
        public Heading Heading { get; private set; }

        public readonly IBoundedRegion _region;
        public readonly IWarningProvider _warningProvider;

        public Ship(IntVector2 position, Heading heading, IBoundedRegion region, IWarningProvider warningProvider)
        {
            this.Position = position;
            this.Heading = heading;
            this._region = region;
            this._warningProvider = warningProvider;
        }

        public void ExecuteInstructions(string instructions)
        {
            foreach(var instruction in instructions)
            {

            }
        }

        private void moveForward()
        {
            this.Position = this.Position + Heading.GetTranslation();
        }

        public enum ShipState
        {
            OK, LOST
        }

    }
}
