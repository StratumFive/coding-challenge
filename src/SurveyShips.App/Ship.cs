using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;

namespace SurveyShips.App
{
    /*
    		        N

        W           			E

		            S
     */
    /// <summary>
    /// Compass Orientation values.
    /// </summary>
    public enum Orientation
    {
        North,
        East,
        South,
        West,
        Unknown
    }

    /// <summary>
    /// Ship object.
    /// </summary>
    public class Ship
    {
        public Point CurrentPosition {get;private set;} 
        public Point Grid {get;private set;}
        public Orientation CurrentOrientation {get;private set;}

        public List<Point> LostShipCoordinates {get;set;}


        public Ship(int startX, int startY, char startOrientation)
        {
            this.CurrentPosition = new Point(startX, startY);
            this.CurrentOrientation = ConvertCharToOrientation(startOrientation);
            this.LostShipCoordinates = new List<Point>();
        }

        /// <summary>
        /// Process the instructions to move and turn a ship.
        /// </summary>
        /// <param name="instructions"></param>
        /// <exception><see cref="ArgumentNullException"/></exception>
        /// <exception><see cref="ArgumentException"/></exception>
        public void ProcessInstructions(string instructions, bool debug = false)
        {
            if(string.IsNullOrEmpty(instructions))
                throw new ArgumentNullException(nameof(instructions));
            
            if(instructions.Length > 99)
                throw new ArgumentException("Instructions need to be less than 100 characters in length ");
            
            for(int i=0;i < instructions.Length;i++)
            {
                switch(char.ToLower(instructions[i]))
                {
                    case 'l':
                        TurnLeft();
                        break;
                    case 'r':
                        TurnRight();
                        break;
                    case 'f':
                        MoveForward();
                        break;
                    default:
                        break;
                }
                if(debug)
                {
                    Console.WriteLine($"Debug: Instruction {instructions[i]} {this.CurrentPosition.X}, {this.CurrentPosition.Y} with a position");
                }            
            }
            
        }

        /// <summary>
        /// Turn orientation of the ship by 90 Degrees to the left.
        /// </summary>
        private void TurnLeft()
        {
            switch(CurrentOrientation)
            {
                case Orientation.North:
                    this.CurrentOrientation = Orientation.West;
                    break;
                case Orientation.East:
                    this.CurrentOrientation = Orientation.North;
                    break;
                case Orientation.South:
                    this.CurrentOrientation = Orientation.East;
                    break;
                case Orientation.West:
                    this.CurrentOrientation = Orientation.South;
                    break;
                default:
                    break;
            }
        }

        /// <summary>
        /// Turn orientation of the ship by 90 Degrees to the left.
        /// </summary>
        private void TurnRight()
        {
            switch(CurrentOrientation)
            {
                case Orientation.North:
                    this.CurrentOrientation = Orientation.East;
                    break;
                case Orientation.East:
                    this.CurrentOrientation = Orientation.South;
                    break;
                case Orientation.South:
                    this.CurrentOrientation = Orientation.West;
                    break;
                case Orientation.West:
                    this.CurrentOrientation = Orientation.North;
                    break;
                default:
                    break;
            }
        }

        /// <summary>
        /// Move Forward one grid point.
        /// </summary>
        private void MoveForward()
        {
             switch(CurrentOrientation)
            {
                case Orientation.North:
                    if(this.LostShipCoordinates.Where(i => i.X ==  CurrentPosition.X && (i.Y == CurrentPosition.Y  + 1))
                                               .Count() > 0)
                        break;
                    this.CurrentPosition = new Point(CurrentPosition.X, CurrentPosition.Y + 1);
                    break;
                case Orientation.East:
                    if(this.LostShipCoordinates.Where(i => i.X + 1 ==  CurrentPosition.X && (i.Y == CurrentPosition.Y))
                                               .Count() > 0)
                        break;
                    this.CurrentPosition = new Point(CurrentPosition.X + 1, CurrentPosition.Y);
                    break;
                case Orientation.South:
                    if(this.LostShipCoordinates.Where(i => i.X ==  CurrentPosition.X && (i.Y == CurrentPosition.Y  - 1))
                                               .Count() > 0)
                        break;
                    this.CurrentPosition = new Point(CurrentPosition.X, CurrentPosition.Y - 1);
                    break;
                case Orientation.West:
                    if(this.LostShipCoordinates.Where(i => i.X - 1 ==  CurrentPosition.X && (i.Y == CurrentPosition.Y))
                                               .Count() > 0)
                        break;
                    this.CurrentPosition = new Point(CurrentPosition.X - 1, CurrentPosition.Y);
                    break;
                default:
                    break;
            }           
        }
        
        /// <summary>
        /// Quick method to return a instruction orientation to <see cref="Orientation.cs" />
        /// </summary>
        /// <param name="orientationCode"></param>
        /// <returns>Orientation</returns>
        private Orientation ConvertCharToOrientation(char orientationCode)
        {
            if(orientationCode == 'N' || orientationCode == 'n')
            {
                return Orientation.North;
            } else if(orientationCode == 'E' || orientationCode == 'e')
            {
                return Orientation.East;
            } else if(orientationCode == 'S' || orientationCode == 's')
            {
                return Orientation.South;
            } else if(orientationCode == 'W' || orientationCode == 'w')
            {
                return Orientation.West;
            }

            return Orientation.Unknown;
        }
    }
}