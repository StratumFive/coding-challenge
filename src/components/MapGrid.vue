<template>
  <div id="mapGrid" ref="mapGrid" class="map-grid-container">
    <div
		v-for="r in rows"
		:key="r"
		class="rowCSS"
	>
		<div 
			v-for="c in columns"
			:key="c"
			class="cellCSS"
			:style="cellSize"
			:id="c"
		>
			<i
				v-if="cellHasShipInCoordinates(r, c )"
				class="fas fa-ship has-ship-icon" :class="iconSize">
			</i>
			<i
				v-else
				class="fas fa-ship" :class="iconSize">
			</i>
		</div>
    </div>
  </div>
</template>

<script>
import { mapGetters } from "vuex"

export default {
	name: "MapGrid",
	
	props: {
		rows: {
			type     : Number,
			required : true
		},

		columns: {
			type     : Number,
			required : true
		}
	},

	computed: {
		...mapGetters({
			hasShipInCoordinates: "hasShipInCoordinates",
			hasWarningInCoordinates: "hasWarningInCoordinates",
		}),

		calculateWidth() {
			return (this.$refs.mapGrid && this.$refs.mapGrid.clientWidth) ? 
				this.$refs.mapGrid.clientWidth / (this.columns + 2)
				: (950 / (this.columns + 2) )
		},

		cellSize() {
			let size = this.calculateWidth
			let margin = "10px"
			let padding = "30px"
			if (this.calculateWidth >= 220) size = this.calculateWidth - 30
			else if (this.calculateWidth >= 180) size = this.calculateWidth - 20
			else if (this.calculateWidth >= 130) size = this.calculateWidth - 10
			else if (this.calculateWidth >= 90) {
				size = this.calculateWidth - 8
				padding = "20px"
			} else if (this.calculateWidth >= 50) {
				size = this.calculateWidth - 4
				margin = "5px"
				padding = "15px"
			} else if (this.calculateWidth >= 30) {
				size = this.calculateWidth
				margin = "2px"
				padding = "10px"
			} else if (this.calculateWidth >= 10) {
				size = this.calculateWidth + 1
				margin = "2px"
				padding = "3px"
			} else {
				size = this.calculateWidth + 2
				margin = "2px"
				padding = "5px"
			}

			return {
				width: `${size}px`,
				height: `${size}px`,
				margin,
				padding
			}
		},

		iconSize() {
			if (this.calculateWidth >= 220) return "fa-7x"
			if (this.calculateWidth >= 150) return "fa-5x"
			if (this.calculateWidth >= 90) return "fa-3x"
			if (this.calculateWidth >= 70) return "fa-2x"
			if (this.calculateWidth >= 50) return "fa-lg"
			if (this.calculateWidth >= 30) return "fa-sm"
			return "fa-xs"
		}
	},

	methods: {
		cellHasShipInCoordinates(row, column) {
			const x = column - 1
			const y = this.rows - row
			return this.hasShipInCoordinates( { x, y })
		}
	}
}
</script>

<style lang="scss" >
	.map-grid-container {
		display: flex;
		flex-direction: column;
		height: 100%;
		border-radius: 10px;
		margin-top: 0;
	}

	.rowCSS {
		display: flex !important;
	}

	.cellCSS {
		display: flex;
		flex-direction: row;
		justify-content: center;
		align-content: center;
		background-color: #0077BE !important;
		padding: 5px;
		margin: 2px;
		border-radius: 5px;
		color: lightgray;
	}

	.has-ship-icon {
		color: darkgoldenrod !important;
	}

	.has-warning-icon {
		color: lightsalmon;
	}
</style>