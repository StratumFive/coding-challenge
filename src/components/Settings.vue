<template>
	<v-form
		ref="form"
		v-model="valid"
		lazy-validation
		class="settings-container"
	>
		<v-row>
			<v-col
				cols="12"
				md="6"
			>
				<v-text-field
					v-model="limitX"
					:counter="2"
					:rules="limitRules"
					label="Limit X"
					required
				></v-text-field>
			</v-col>
			<v-col
				cols="12"
				md="6"
			>
				<v-text-field
					v-model="limitY"
					:counter="2"
					:rules="limitRules"
					label="Limit Y"
					required
				></v-text-field>
			</v-col>
		</v-row>
		
		<v-btn
			color="error"
			class="mr-2"
			@click="reset"
        >
			<v-icon small>fa-refresh </v-icon>
		</v-btn>

		<v-btn
			color="warning"
			@click="setWorldLimits"
		>
			New World
				<v-icon right>
					fas fa-globe sm
				</v-icon>
		</v-btn>
		
		<v-alert
			:value="showMessage"
			border="top"
			class="mt-3"
			colored-border
			:type="alertType"
			elevation="2"
			transition="scale-transition"
			dismissible
			>	{{ results }}
		</v-alert>
  </v-form>
</template>

<script>
import { mapActions } from "vuex"

export default {
	name: "Settings",
	
    data: () => ({
		valid: true,
		limitX: 5,
		limitY: 3,
		limitRules: [
			v => !!v || "Limit X is required",
			v => (v && v.length <= 2) || "The limits should have at most 2 characters",
		],
		alertType: "success",
		showMessage: false,
		results: "",
    }),

    methods: {
		...mapActions({
			createWorld: "createWorld",
		}),

		reset () {
			this.$refs.form.reset()
			this.$emit("resetWorld")
		},

		async setWorldLimits() {
			try {
				await this.createWorld({ 
					limitX: Number(this.limitX), 
					limitY: Number(this.limitY)
				})
				this.results = "The world has been created!|"
				this.showMessage = true
				this.alertType = "success"
				this.$emit("createdWorld", { limitX: this.limitX, limitY: this.limitY })
			} catch(error) {
				this.results = error
				this.showMessage = true
				this.alertType = "error"
			}
		},
    },
  }
</script>

<style lang="scss" scoped>
	.settings-container {
		max-width: 250px;
		margin-left: 1rem;
	}
</style>