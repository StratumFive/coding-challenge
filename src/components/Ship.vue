<template>
	<v-form
		ref="form"
		v-model="valid"
		lazy-validation
		class="ship-container"
	>
		<v-row>
			<v-col
				cols="12"
				md="6"
			>
				<v-text-field
					v-model="coordinateX"
					:counter="2"
					:rules="coordinatesRules"
					label="X Coordinate"
				></v-text-field>
			</v-col>
			<v-col
				cols="12"
				md="6"
			>
				<v-text-field
					v-model="coordinateY"
					:counter="2"
					:rules="coordinatesRules"
					label="Y Coordinate"
				></v-text-field>
			</v-col>
		</v-row>
		
		<v-select
			v-model="currentOrientation"
			:items="orientations"
			label="Orientation"
			item-text="label"
			item-value="value"
        ></v-select>
		
		<v-btn
			color="submit"
			class="mr-4"
			@click="addInstruction('L')"
		>L</v-btn>
		<v-btn
			color="submit"
			class="mr-4"
			@click="addInstruction('R')"
		>R</v-btn>
		<v-btn
			color="submit"
			class="mr-4"
			@click="addInstruction('F')"
		>F</v-btn>

		<v-textarea
			clearable
			clear-icon="mdi-close-circle"
			label="Add Insrtuctions"
			v-model="instructions"
			class="mt-2"
			rows="2"
			row-height="20"
		></v-textarea>

		<v-btn
			color="error"
			class="mr-2 mb-4"
			@click="reset"
        >
			<v-icon small>fa-refresh </v-icon>
		</v-btn>


		<v-btn
			color="warning"
			class="mr-4 mb-4"
			@click="executeInstructions"
		>
			Add Ship
				<v-icon right>
					fas fa-ship sm
				</v-icon>
		</v-btn>
		
		<v-alert
			:value="showMessage"
			border="top"
			class="mt-3"
			colored-border
			:type="alertType"
			elevation="2"
			transition="scale-transition"
			dismissible
			>	{{ results }}
		</v-alert>
  </v-form>
</template>

<script>
import { mapGetters, mapActions } from "vuex"
import { orientation } from "@services/shipUtils"

export default {
	name: "Settings",
	
    data: () => ({
		valid: true,
		coordinateX: 1,
		coordinateY: 1,
		currentOrientation: "",
		instructions: [],
		coordinatesRules: [
			v => !!v || "Required",
			v => (v && v < 50) || "The maximum value is 50"
		],
		showMessage: false,
		alertType: "success",
		results: "",
    }),

	computed: {
		...mapGetters({
			currentShip: "currentShip",
		}),

		orientations() {
			return Object.entries(orientation).map(iter => ({
				label: iter[0],
				value: iter[1]
			}))
		}
	},

    methods: {
		...mapActions({
			moveShip: "moveShip",
		}),

		reset () {
			this.$refs.form.reset()
		},

		addInstruction(instruction) {
			this.instructions.push(instruction)
		},

		async executeInstructions() {
			const ship = {
				x: Number(this.coordinateX),
				y: Number(this.coordinateY),
				orientation: this.currentOrientation,
				status: "OK"
			}
			let instructions
			if (Array.isArray(this.instructions)) instructions = this.instructions
			else instructions = Array.from(this.instructions)
			try {
				await this.moveShip({ ship, instructions })
				this.results = `Coordinates: ${this.currentShip.x} ${this.currentShip.y}
							- Orientation: ${this.currentShip.orientation}
							Status: ${this.currentShip.status}`
				this.showMessage = true
				this.alertType = "success"

				// reset the form
				this.$refs.form.reset()
				this.$emit("newShip")
			} catch(error) {
				this.results = error
				this.showMessage = true
				this.alertType = "error"
				console.log("FULL ERR", error)
			}

		}
    },
  }
</script>

<style lang="scss" scoped>
	.ship-container {
		max-width: 250px;
		margin-left: 1rem;
	}
</style>