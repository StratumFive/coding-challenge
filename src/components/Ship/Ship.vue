<template>
  <div class="flex flex-col">
    <svg :height="this.yMax * 40" class="border border-black rounded w-full">
      <g v-for="y in yMax + 1" :key="y">
        <circle
          v-for="x in xMax + 1"
          :key="x"
          :cx="`${(100 / (xMax + 2)) * x}%`"
          :cy="`${(100 / (yMax + 2)) * y}%`"
          r="4"
          class="fill-current text-black"
        />
      </g>
      <ShipIcon
        :orientation="orientation"
        :x="x"
        :y="y"
        class="w-8"
        :class="isLost ? { 'text-red-500': true } : { 'text-green-500': true }"
      />
    </svg>
    <div class="mt-8 flex justify-between">
      <Button
        @click.native="reset()"
        class="bg-teal-200 border border-teal-700 hover:bg-teal-100 w-32"
      >
        Reset
      </Button>
      <Button
        @click.native="followNextInstruction"
        class="bg-teal-200 border border-teal-700 hover:bg-teal-100 w-32"
      >
        Next
      </Button>
    </div>
  </div>
</template>

<script>
import ShipIcon from './ShipIcon'
import Button from '@/components/Form/Button'
import { LEFT_TURNS, MOVES, RIGHT_TURNS } from '@/constants/instructions'
import { mapGetters, mapState } from 'vuex'

export default {
  name: 'Ship',
  components: {
    Button,
    ShipIcon,
  },
  props: {
    id: {
      required: true,
      type: Number,
    },
    initialOrientation: {
      required: true,
      type: String,
      validator(orientation) {
        return ['N', 'E', 'S', 'W'].includes(orientation)
      },
    },
    initialX: {
      required: true,
      type: Number,
      validator(x) {
        return x <= 50
      },
    },
    initialY: {
      required: true,
      type: Number,
      validator(y) {
        return y <= 50
      },
    },
    instructions: {
      required: true,
      type: String,
      validator(instructions) {
        return instructions.length < 100
      },
    },
  },
  data() {
    return {
      instructionsFollowed: 0,
      orientation: this.initialOrientation,
      x: this.initialX,
      y: this.initialY,
    }
  },
  computed: {
    // Returns coordinates within the frame of the grid, even if this ship fell off
    boundedCoordinates() {
      return {
        x: Math.max(0, Math.min(this.x, this.xMax)),
        y: Math.max(0, Math.min(this.y, this.yMax)),
      }
    },
    isCurrentPositionDangerous() {
      return this.getLostShips(this.id).some(
        ({ orientation, x, y }) =>
          orientation === this.orientation && x === this.x && y === this.y
      )
    },
    isLost() {
      return (
        this.x < 0 || this.x > this.xMax || this.y < 0 || this.y > this.yMax
      )
    },
    ...mapState(['xMax', 'yMax']),
    ...mapGetters(['getLostShips']),
  },
  created() {
    for (const instruction of this.instructions) {
      this.followInstruction(instruction)
    }

    this.$emit('signal-final-position', {
      ship: {
        id: this.id,
        isLost: this.isLost,
        orientation: this.orientation,
        ...this.boundedCoordinates,
      },
    })

    this.reset()
  },
  methods: {
    followInstruction(instruction) {
      if (this.isLost) {
        return
      }

      if (instruction === 'F') {
        this.move()
      } else if (instruction === 'L') {
        this.orientation = LEFT_TURNS[this.orientation]
      } else if (instruction === 'R') {
        this.orientation = RIGHT_TURNS[this.orientation]
      }
    },
    followNextInstruction() {
      this.followInstruction(this.instructions[this.instructionsFollowed])
      this.instructionsFollowed++
    },
    move() {
      if (this.isCurrentPositionDangerous) {
        return
      }

      const [xDiff, yDiff] = MOVES[this.orientation]
      this.x += xDiff
      this.y += yDiff
    },
    reset() {
      this.instructionsFollowed = 0
      this.orientation = this.initialOrientation
      this.x = this.initialX
      this.y = this.initialY
    },
  },
}
</script>
