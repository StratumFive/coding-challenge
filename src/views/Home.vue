<template>
	<v-card class="mt-0 pt-0">
		<div name="mainHomeComponent" class="main-home-component">
			<div name="worldSettingsComponentContainer" class="components">
				<v-card-title class="pb-0">World Settings</v-card-title>
				<settings @createdWorld="loadGrid" @resetWorld="resetWorld" />
				
				<v-divider class="mt-4" />
				<v-card-title class="pt-5 pb-0">Add Ship</v-card-title>
				<ship @newShip="updateGrid" />
			</div>
			<div name="mapGridComponentContainer" class="components">
				<map-grid
					:key="gridComponentKey"
					v-show="showMapGrid"
					:rows="limitY"
					:columns="limitX"
				/>
			</div>
		</div>
	</v-card>
</template>

<script>
import { mapGetters, mapActions } from "vuex"
import Settings from "@components/Settings"
import Ship from "@components/Ship"
import MapGrid from "@components/MapGrid"

export default {
	name: "Home",

	components: {
		Settings,
		Ship,
		MapGrid,
	},

    data: () => ({
		showMapGrid: false,
		limitX: 0,
		limitY: 0,
		gridComponentKey: 0
	}),
	
	computed: {
		...mapGetters({
			currentShip: "currentShip",
		}),
	},

	methods: {
		...mapActions({
			createWorld: "createWorld",
			moveShip: "moveShip",
		}),

		/**
		 * * [ loadGrid: This method is triggered from the child component Settings once the user has 
		 * 		successfully created a new world. With the limits already set by the user, this method enforces a new load of 
		 * 		the MapGrid component with the new world just created. ]
		 * @param {[Object]} limits	[ represents maximum x and y coordinate for any ship (to create  new world) ]
		 */
		loadGrid(limits) {
			if (limits && limits.limitX > 0 && limits.limitY > 0) {
				this.limitX = parseInt(limits.limitX, 10)
				this.limitY = parseInt(limits.limitY, 10)
				this.gridComponentKey += 1
				this.showMapGrid = true
			}
		},

		/**
		 * * [ resetWorld: This method is triggered from the child component Settings once the user has 
		 * 		successfully reset the active world. This method enforces to hide the map component. ]
		 */
		resetWorld(){
			this.showMapGrid = false
			this.limitX = 0
			this.limitY = 0
		},

		/**
		 * * [ updateGrid: This method is triggered from the child Ship component once the user has 
		 * 		successfully added a new ship instruction. This method enforces to hide the map component
		 * 		to show the new ship on it. ]
		 */
		updateGrid(){
			this.gridComponentKey += 1
		},
	}

};
</script>

<style lang="scss" >
	.main-home-component {
		display: grid;
		grid-template-columns: 20% 80%;
	}

	.components {
		margin-bottom: 2rem;
	}
</style>