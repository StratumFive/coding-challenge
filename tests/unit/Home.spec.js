import { shallowMount, createLocalVue } from "@vue/test-utils"
import Vue from "vue"
import Vuex from "vuex"
import Vuetify from "vuetify"
import Home from "@views/Home"
import Settings from "@components/Settings"
import Ship from "@components/Ship"
import MapGrid from "@components/MapGrid"
import shipStore from "@store/index"

// import router from "@/router"
Vue.use(Vuetify)
const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(Vuetify)

let wrapper
let state
let store
let vuetify
let defaultShip

afterEach(() => {
	wrapper.destroy()
})

describe("testing Home Template", () => {
	beforeEach(() => initialization())

	it("should render the page and match the snapshot", () => {
		// tests that the wrapper was loadded properly
		expect(wrapper.exists()).toBe(true)
		expect(wrapper).toMatchSnapshot()
	})

	it("should render one main div component with the proper class", () => {
		expect(wrapper.find("div[name=\"mainHomeComponent\"]").exists()).toBe(true)
		expect(wrapper.find("div[name=\"mainHomeComponent\"]").classes()).toContain("main-home-component")
	})

	it("should render 2 container div components with the proper classes", () => {
		expect(wrapper.find("div[name=\"worldSettingsComponentContainer\"]").exists()).toBe(true)
		expect(wrapper.find("div[name=\"worldSettingsComponentContainer\"]").classes()).toContain("components")
		expect(wrapper.find("div[name=\"mapGridComponentContainer\"]").exists()).toBe(true)
		expect(wrapper.find("div[name=\"mapGridComponentContainer\"]").classes()).toContain("components")
	})

	it("should render one Settings component", () => 
		expect(wrapper.findComponent(Settings).exists()).toBe(true))

	it("should render one Ship component", () => 
		expect(wrapper.findComponent(Ship).exists()).toBe(true))

	it("should render one MapGrid component with the right props", () => {
		expect(wrapper.findAllComponents(MapGrid).length).toBe(1)
		const mapGridComponent = wrapper.findAllComponents(MapGrid).at(0)
		expect(mapGridComponent.props().rows).toBe(0)
		expect(mapGridComponent.props().columns).toBe(0)
	})
})

describe("testing Home Computed properties", () => {
	beforeEach(() => initialization())

	describe("testing currentShip computed", () => {
		it("should return the default empty ship, when initialized", () =>
			expect(wrapper.vm.currentShip).toEqual(defaultShip))

		it("should return a real ship, when there is one on the state", () => {
			// setting the store to test
			const movedShip = {
				x: 10,
				y: 20,
				orientation: "W",
				status: "OK"
			}
			shipStore.mutations.SET_MOVED_SHIP(state, movedShip)

			// asserting that the getter returns the proper ship
			expect(wrapper.vm.currentShip).toEqual(movedShip)
		})
	})
})

describe("testing Home Methods", () => {
	beforeEach(() => initialization())

	describe("testing loadGrid method", () => {
		it("should NOT change the limits or the showMapGrid property, if the limits are undefined", async() => {
			// setting default values to test that after the method, they are still the same
			wrapper.vm.limitX = 10
			wrapper.vm.limitY = 12
			wrapper.vm.showMapGrid = false

			// trigger the method and wait
			wrapper.vm.loadGrid()
			await wrapper.vm.$nextTick()

			// assert
			expect(wrapper.vm.limitX).toBe(10)
			expect(wrapper.vm.limitY).toBe(12)
			expect(wrapper.vm.showMapGrid).toBe(false)
		})

		it("should NOT change the limits or the showMapGrid property, if the X limit is not correct", async() => {
			// setting default values to test that after the method, they are still the same
			wrapper.vm.limitX = 10
			wrapper.vm.limitY = 12
			wrapper.vm.showMapGrid = false

			// trigger the method and wait
			wrapper.vm.loadGrid({ limitX: -1, limitY: 10 })
			await wrapper.vm.$nextTick()

			// assert
			expect(wrapper.vm.limitX).toBe(10)
			expect(wrapper.vm.limitY).toBe(12)
			expect(wrapper.vm.showMapGrid).toBe(false)
		})

		it("should NOT change the limits or the showMapGrid property, if the Y limit is not correct", async() => {
			// setting default values to test that after the method, they are still the same
			wrapper.vm.limitX = 10
			wrapper.vm.limitY = 12
			wrapper.vm.showMapGrid = false

			// trigger the method and wait
			wrapper.vm.loadGrid({ limitX: 10, limitY: 0 })
			await wrapper.vm.$nextTick()

			// assert
			expect(wrapper.vm.limitX).toBe(10)
			expect(wrapper.vm.limitY).toBe(12)
			expect(wrapper.vm.showMapGrid).toBe(false)
		})

		it("should set the new limits, set the showMapGrid property to true and update the mapGridKey, if the limits are correct", async() => {
			// setting default values to test that after the method, they are still the same
			wrapper.vm.limitX = 10
			wrapper.vm.limitY = 12
			wrapper.vm.showMapGrid = false
			wrapper.vm.gridComponentKey = 1

			// trigger the method and wait
			wrapper.vm.loadGrid({ limitX: 40, limitY: 16 })
			await wrapper.vm.$nextTick()

			// assert
			expect(wrapper.vm.limitX).toBe(40)
			expect(wrapper.vm.limitY).toBe(16)
			expect(wrapper.vm.showMapGrid).toBe(true)
			expect(wrapper.vm.gridComponentKey).toBe(2)
		})
	})

	describe("testing resetWorld method", () => {
		it("should set the new limits to zero, set the showMapGrid property to false and update the mapGridKey", async() => {
			// setting default values to test that after the method, they are still the same
			wrapper.vm.limitX = 10
			wrapper.vm.limitY = 12
			wrapper.vm.showMapGrid = false

			// trigger the method and wait
			wrapper.vm.resetWorld()
			await wrapper.vm.$nextTick()

			// assert the reset
			expect(wrapper.vm.limitX).toBe(0)
			expect(wrapper.vm.limitY).toBe(0)
			expect(wrapper.vm.showMapGrid).toBe(false)
		})
	})

	describe("testing updateGrid method", () => {
		it("should set the new limits to zero, set the showMapGrid property to false and update the mapGridKey", async() => {
			wrapper.vm.gridComponentKey = 19

			// trigger the method and wait
			wrapper.vm.updateGrid()
			await wrapper.vm.$nextTick()

			expect(wrapper.vm.gridComponentKey).toBe(20)
		})
	})
})

/**
 * [initialization is a function that defines all the boiler plate code to create the Home
 * component with mocked data to be tested.
 * This function also creates the state used on the instance. ]
 */
async function initialization(newWorld, limitX=0, limitY=0) {
	vuetify = new Vuetify()

	defaultShip = {
		x: -1,
		y: -1,
		orientation: "",
		status: "",
	}

	state = { 
		limitX,
		limitY,
		world: (newWorld)? newWorld : [],
		currentShip: { ...defaultShip },
		warnings: []
	}

	store = new Vuex.Store({
		state,
		getters   : shipStore.getters,
		actions   : shipStore.actions,
		mutations : shipStore.mutations
	})

	// mount the wrapper
	wrapper = shallowMount(Home, {
		localVue,
		store,
		vuetify,
	})
	await wrapper.vm.$nextTick()
}