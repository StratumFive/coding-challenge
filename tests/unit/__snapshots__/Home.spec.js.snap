// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing Home Template should render the page and match the snapshot 1`] = `
<v-card-stub loaderheight="4" tag="div" class="mt-0 pt-0">
  <div name="mainHomeComponent" class="main-home-component">
    <div name="worldSettingsComponentContainer" class="components">
      <v-card-title-stub class="pb-0">World Settings</v-card-title-stub>
      <settings-stub></settings-stub>
      <v-divider-stub class="mt-4"></v-divider-stub>
      <v-card-title-stub class="pt-5 pb-0">Add Ship</v-card-title-stub>
      <ship-stub></ship-stub>
    </div>
    <div name="mapGridComponentContainer" class="components">
      <map-grid-stub rows="0" columns="0" style="display: none;"></map-grid-stub>
    </div>
  </div>
</v-card-stub>
`;
